### PR TITLE
Added support for multi threading.

### PR DESCRIPTION
**Pull request by weskubo-cgi**
Fixed issues with static JsonParsers.
Made HandshakeUtil methods static.
Renamed HandshakeServerTest to HandshakeUtilTest to better reflect it's purpose.

### CODE REVIEW

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java [hunk 1]
```diff
@@ -26,140 +23,63 @@
...
+						logger.info("{} - Accepting connection attempt from IP Address: {}", LoggingUtil.getMethodName(),
+								connectionSocket.getInetAddress().getCanonicalHostName());
+						
+						ConnectionHandler handler = new ConnectionHandler(producer, connectionSocket, properties.getSocketReadSleepTime(), properties.getMaxSocketReadTries());
+						
+						Integer activeCount = executor.getActiveCount();
+						
+						logger.debug("{} threads are active", activeCount);
+						if (activeCount == properties.getThreadPoolSize()) {
+							logger.info("Max Thread Pool Size reached. Waiting for available thread");
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java:67]: There is no logic implemented to wait. After logging, control will go to next step.

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java [hunk 2]
```diff
@@ -169,179 +89,11 @@ public void run() {
 							logger.warn("ServerSocket could not be closed");
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java:89]: (Not part of this commit), Should we should update the logs as per new syntax?

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ServerProperties.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class ServerProperties {
+
+	private Integer serverSocket;
+	private Integer socketReadSleepTime;
+	private Integer maxSocketReadTries;
+	private Integer threadPoolSize;
+
+	public ServerProperties(Integer serverSocket, Integer socketReadSleepTime, Integer maxSocketReadTries, Integer threadPoolSize) {
+		super();
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ServerProperties.java]: Unnecessary line of code to call super().

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+	}
+	
+	/**
+	 * Reads client message and sends request to ESB
+	 * @param socketInput
+	 * @param socketOutput
+	 * @throws IOException
+	 * @throws InterruptedException
+	 */
+	private void performTransaction(BufferedInputStream socketInput, BufferedOutputStream socketOutput)
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java:177]: (Not part of this commit) This method is doing too many actions:
// read SI segment
// read dtsegment header
// Look for the start of data
// read dtsegment data
// Send received message to server
// Write Response back to POS
and this is private method.. so no JUnit for this one.
May be we should create a story of looking into breaking these tasks into smaller methods and add Junits for those?

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+			// Depending on when the handshake process fails the ret_code might actually return HNET_RTRN_SUCCESS
+			// If this is the case use a HNET_RTRN_INVALIDFORMATERROR instead as it is likely a formatting error
+			if (ret_code.contentEquals(MessageUtil.HNET_RTRN_SUCCESS)) {
+				ret_code = MessageUtil.HNET_RTRN_INVALIDFORMATERROR;
+			}
+			hnSecureResponse = ErrorBuilder.buildDefaultErrorMessage(ret_code);
+			try {
+				sendResponse(socketOutput, hnSecureResponse);
+			} catch (IOException ioe) {
+				logger.error(ioe.getMessage());
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java:108]: (not part of this commit) Syntax is not matching as other logs. Same comment for other logger.error(...)

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+import org.apache.camel.ProducerTemplate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ca.bc.gov.hlth.hnclientv2.error.ErrorBuilder;
+import ca.bc.gov.hlth.hnclientv2.error.MessageUtil;
+import ca.bc.gov.hlth.hncommon.util.LoggingUtil;
+import io.netty.util.internal.StringUtil;
+
+public class ConnectionHandler implements Callable<Void> {
```
Comment (pankaj-cgi) [inline hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ConnectionHandler.java:20]: (Not part of this commit) Wondering if we have enough junit for business logic in this class or we need to create a story to add junits?

**Review by pankaj-cgi — COMMENTED** [review]
Changes in this commit look all good. Just one question/comment for changes in this commit.
Rest all comments are related to older code. May be we can create a bug/story to work on older code if required.

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java (hunk 1 above)
Comment (weskubo-cgi) [reply hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/HandshakeServer.java:67]: The logging is strictly informational. The ThreadPoolExecutor will handle the submit() call with the next available thread.

File: hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ServerProperties.java (hunk 3 above)
Comment (weskubo-cgi) [reply hnclient-v2/src/main/java/ca/bc/gov/hlth/hnclientv2/handshakeserver/ServerProperties.java]: Eclipse code gen :). I'll remove.
